Add jaeger exporter

daniel-abramov opened the pull request.

daniel-abramov committed:
telemetry: add trace provider and jaeger support

daniel-abramov committed:
telemetry: add configuration possibilities

daniel-abramov committed:
telemetry: activate telemetry if config provided

daniel-abramov committed:
telemetry: discover/detect container in resource

daniel-abramov committed:
configs: add examples for the telemetry configs

dbkr reviewed (commented):
Otherwise looks good

daniel-abramov committed:
telemetry: do not use global tracer

Replace it by requesting the tracer each time we create a new telemetry
span. This is not super inefficient, but avoids the usage of a global
namespace.

daniel-abramov commented on pkg/telemetry/setup.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+// Global internal tracer that generates spans.
+var tracer = otel.Tracer("undefined")
```
Initially, the idea was that I stored the tracer in a private global variable so that those who want to use `NewTelemetry()` don't need to supply an additional parameter called "tracer" (that they would need to either create or get it from the parent). But then I figured that `otel` actually sets a global tracer whenever we initialize and explicitly set it in the set-up function, i.e. we could access this tracer any time by just calling `otel.GetTracingProvider().Tracer()`, so there is actually no need to have an additional global variable in our case (also global variables are bad 🙂), so I removed this part completely! Thanks for pointing it out!

daniel-abramov commented on pkg/telemetry/setup.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// Create a new Jaeger exporter.
+	exp, err := NewJaegerExporter(config.JaegerURL)
```
Interesting. I've totally forgotten about the OTLP protocol. I've just checked how to add the support of OTLP in Go and noticed that there are 2 ways to send things over OTLP:

- HTTP with protobuf encoded content
- gRPC

I've checked the Jaeger setup from the branch you shared with me and if I get it right, our logger instance supports both. Does it mean I can choose any of these (depending on which one is simpler to set up) for the SFU? Do you know if I need to attach any credentials to it? I've noticed that there is a possibility to pass the token:

```go
      otlptracegrpc.WithHeaders(map[string]string{
          "Authorization": "<ADD YOUR TOKEN HERE>",
      }))
```

dbkr replied:
The docker setup is the jaeger all-in-one which does support a wide set of protocols, but if we're able to use OTLP over HTTP then this will be consistent with element call. I don't think we'll need any access tokens.

daniel-abramov replied:
Makes sense! Added support of OTLP (the option to use Jaeger remains, the user can choose the exporter by configuring one or another one, i.e. I use the jaeger URL locally for the tests, but I've just tested it with the OTLP exporter and it also works fine as long as the local instance of Jaeger configures OTLP).

daniel-abramov committed:
telemetry: add support of OTLP protocol
